src/main/scala/ai/privado/languageEngine/csharp/tagger/collection/CollectionTagger.scala

```diff
@@ -0,0 +1,139 @@
+ *
+ * Privado is an open source static code analysis tool to discover data flows in the code.
+ * Copyright (C) 2022 Privado, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * For more information, contact support@privado.ai
+ *
+ */
+
+package ai.privado.languageEngine.csharp.tagger.collection
+
+import ai.privado.cache.RuleCache
+import ai.privado.languageEngine.java.tagger.collection.CollectionUtility
+import ai.privado.model.{Constants, RuleInfo}
+import ai.privado.tagger.PrivadoParallelCpgPass
+import io.shiftleft.codepropertygraph.generated.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.{Annotation, Method}
+import io.shiftleft.semanticcpg.language.*
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+
+class CollectionTagger(cpg: Cpg, ruleCache: RuleCache) extends PrivadoParallelCpgPass[RuleInfo](cpg) {
+  protected val methodUrlMap: mutable.HashMap[Long, String] = mutable.HashMap[Long, String]()
+  protected val classUrlMap: mutable.HashMap[Long, String]  = mutable.HashMap[Long, String]()
+  private val logger                                        = LoggerFactory.getLogger(this.getClass)
+
+  override def generateParts(): Array[RuleInfo] =
+    ruleCache.getRule.collections.filter(_.catLevelTwo == Constants.annotations).toArray
+
+  override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
+    classUrlMap.addAll(collectAnnotatedUrlsFromClasses(ruleInfo.combinedRulePattern))
+    val (methodUrls, collectionMethodsCache) = collectAnnotatedUrlsFromMethods(ruleInfo.combinedRulePattern)
+    methodUrlMap.addAll(methodUrls)
+    tagSources(builder, ruleInfo, collectionMethodsCache)
+  }
+
+  protected def tagSources(
+    builder: DiffGraphBuilder,
+    ruleInfo: RuleInfo,
+    collectionMethodsCache: List[Method]
+  ): Unit = {
+    CollectionUtility.tagDirectSources(
+      builder,
+      collectionMethodsCache,
+      ruleCache.getRule.sources,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+
+    CollectionUtility.tagDerivedSources(
+      cpg,
+      builder,
+      collectionMethodsCache,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+  }
+
+  private def getUrlFromAnnotationsCode(annotation: Annotation): String = {
```

Comment:
When calling this function, we know before hand if this annotation is placed on a method or typeDecl, that information can be passed to this function and that node can be used for below queries, by this quering on cpg can be limited

Reply:
We do need both queries to run on types as well as methods given the different ways in which routes can be defined, but I think I got another idea from this. Will implement this 👍